### PR TITLE
Feat/write operations

### PR DESCRIPTION
**Pull request by greged93**
Time spent on this PR: 3 days

Resolves: #728 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?
All of the write operations have been migrated to use the provider.

# Does this introduce a breaking change?

- [x] Yes
- [ ] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741)

### CODE REVIEW

**Review by Eikix — CHANGES_REQUESTED** [review]
> Reviewed 79 of 79 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741)* status: all files reviewed (commit messages unreviewed), 6 unresolved discussions (waiting on @ClementWalter and @greged93)

___
*[`src/eth_provider/provider.rs` line 383 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTeBjH8_zW_d9seMIP:-NpTeBjH8_zW_d9seMIQ:b-1fhbgh) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L383)):*
> ```XML
>         // Set a high gas limit to make sure the transaction will not fail due to gas.
>         let mut call = call;
>         call.gas = Some(U256::from(*MAX_FEE));
> ```

Does this work? maybe it's a bit more idiomatic



_Suggestion:_
```XML
        let call = CallRequest {..call, gas: Some(U256::from(*MAX_FEE)) } ;
```
___
*[`src/eth_provider/provider.rs` line 405 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTf21N-TRqi_4U2ayX:-NpTf21N-TRqi_4U2ayY:b-hxsl6n) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L405)):*
> ```XML
>         };
> 
>         let block_count = newest_block - oldest_block + U256::from(1);
> ```

why do we need to set it again?
___
*[`src/eth_provider/provider.rs` line 433 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTfTTM5hUhNG2MjFQD:-NpTfTTM5hUhNG2MjFQE:bk0quib) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L433)):*
> ```XML
>         let signer = transaction_signed
>             .recover_signer()
>             .ok_or_else(|| ConversionError::ToStarknetTransactionError("Failed to recover signer".to_string()))?;
> ```

why is this a ToStarknetTransactionError? it's a recovery error
___
*[`src/eth_provider/utils.rs` line 19 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTgDKu7toe_ZLZ3doS:-NpTgDKu7toe_ZLZ3doT:b-a7n9ne) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/utils.rs#L19)):*
> ```XML
>     // Add 2 to the width to account for the 0x prefix.
>     let s = format!("{:#0width$x}", value, width = width + 2);
>     // This can happen because of the LowerHex implementation for Uint,
> ```

what is "This" ? in the "This can happen"
___
*[`src/eth_provider/starknet/kakarot_core.rs` line 64 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTgypa7LfJio2FXZAj:-NpTgypa7LfJio2FXZAk:b-5yzpn1) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/starknet/kakarot_core.rs#L64)):*
> ```XML
>     let nonce = FieldElement::from(transaction.nonce());
> 
>     // Get estimated_fee from Starknet
> ```

we can delete this comment as this is not true anymore
___
*[`src/eth_provider/starknet/kakarot_core.rs` line 95 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTh8BN6ZO1eWTXsyc4:-NpTh8BN6ZO1eWTXsyc5:b-thwmix) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/starknet/kakarot_core.rs#L95)):*
> ```XML
>         FieldElement::ONE,                     // call array length
>         *KAKAROT_ADDRESS,                      // contract address
>         *ETH_SEND_TRANSACTION,                 // selector
> ```

Do we need this selector? For now the EOA only ever calls eth_send_transaction,
I guess it's useful as a reading helper

**Review by greged93 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741)* status: all files reviewed (commit messages unreviewed), 6 unresolved discussions (waiting on @ClementWalter and @Eikix)

___
*[`src/eth_provider/provider.rs` line 383 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTeBjH8_zW_d9seMIP:-NpTjewuC22EBCHMdO2J:bybda6s) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L383)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

Does this work? maybe it's a bit more idiomatic
</blockquote></details>

interesting, did not know you could do that, nice!
___
*[`src/eth_provider/provider.rs` line 405 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTf21N-TRqi_4U2ayX:-NpTjjm6C8Evb7Ik8UYJ:b-s73sme) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L405)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

why do we need to set it again?
</blockquote></details>

let's say you pass block_count = 20 and newest = 10. That means that actually you don't have 20 blocks in that range, oldest will be 0 and you should only have block_count = 11.
___
*[`src/eth_provider/provider.rs` line 433 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTfTTM5hUhNG2MjFQD:-NpTkAfu2bcdi8JhYOi4:beuh9u2) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L433)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

why is this a ToStarknetTransactionError? it's a recovery error
</blockquote></details>

Right but this is used in the process in converting a ethereum transaction to a starknet transaction.
___
*[`src/eth_provider/utils.rs` line 19 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTgDKu7toe_ZLZ3doS:-NpTkIDVAq3s0wC1yLoZ:bvu6gd4) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/utils.rs#L19)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

what is "This" ? in the "This can happen"
</blockquote></details>

s.len() < width. Normally the formatting should pad to a width of 66, but it can be overwritten if the value is 0
___
*[`src/eth_provider/starknet/kakarot_core.rs` line 95 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTh8BN6ZO1eWTXsyc4:-NpTkXPT6rUGTDUqKp2K:b8k6j6b) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/starknet/kakarot_core.rs#L95)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

Do we need this selector? For now the EOA only ever calls eth_send_transaction,
I guess it's useful as a reading helper
</blockquote></details>

I think it doesn't hurt to keep it, makes for better readability

**Comment by Eikix** [comment]
*[`src/eth_provider/utils.rs` line 19 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTgDKu7toe_ZLZ3doS:-NpTnT_P-poBXomrCqj9:b-mznnyt) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/utils.rs#L19)):*
<details><summary><i>Previously, greged93 wrote…</i></summary><blockquote>

s.len() < width. Normally the formatting should pad to a width of 66, but it can be overwritten if the value is 0
</blockquote></details>

I think the comment should be clearer:) in general we should limit the use of "this" , "it" etc. in comments as it might confuse people

**Comment by Eikix** [comment]
*[`src/eth_provider/provider.rs` line 405 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTf21N-TRqi_4U2ayX:-NpTnsEY2D6sOicYYODi:b6e8z4c) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L405)):*
<details><summary><i>Previously, greged93 wrote…</i></summary><blockquote>

let's say you pass block_count = 20 and newest = 10. That means that actually you don't have 20 blocks in that range, oldest will be 0 and you should only have block_count = 11.
</blockquote></details>

can we check if there isn't an early return in other clients? (reth?) if block_count > newest_block return default?

**Comment by Eikix** [comment]
*[`src/eth_provider/provider.rs` line 433 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTfTTM5hUhNG2MjFQD:-NpTo-l7DyI6rN70CLRd:b-birh3c) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L433)):*
<details><summary><i>Previously, greged93 wrote…</i></summary><blockquote>

Right but this is used in the process in converting a ethereum transaction to a starknet transaction.
</blockquote></details>

Alright fair

**Comment by Eikix** [comment]
*[`src/eth_provider/provider.rs` line 411 at r5](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpVI68_3tTQRsX2e5U6:-NpVI68a7AS9ji-LqH6I:b-3qdvn3) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/645f42b230754924571d0b5a582fa9e8a990014b/src/eth_provider/provider.rs#L411)):*
> ```XML
>         // This comes from the rpc spec: `An array of block base fees per gas.
>         // This includes the next block after the newest of the returned range,
>         // because this value can be derived from the newest block. Zeroes are returned for pre-EIP-1559 blocks.`
> ```

this is a bit weird because in the case of Kakarot, we "consider' that basefee := gasPrice and priorityFee := 0 at all times for compatibility.

TL;DR: we should choose for coherence to use the historical block's basefee per gas using the DB.


1. Use the feeder gateway L1 gasPrice, multiply it by a multiple between 1 and 2 (to account for L2 gas price) for each block and consider it to be the block's baseFee. This is what each user will pay to be included in this block, at a fixed cost.
2. This is now the block's baseFee (for the indexer). Each tx that has a maxFeePerGas below this gasprice will be censored.
3. In the fee history, we can use this base fee as the price of the Kakarot L2.


This gas subject is very tricky indeed! 
- because it is two-dimensional
- because we're trying to be compatible with EIP1559 when we're not

Open to challenges.

It can be a separate macro-task / new issue!

**Review by Eikix — COMMENTED** [review]
> Reviewed 1 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741)* status: 71 of 80 files reviewed, 2 unresolved discussions (waiting on @ClementWalter and @greged93)

File: src/test_utils/mongo/mod.rs [hunk 1]
```diff
@@ -84,7 +97,24 @@ pub async fn mock_database() -> Database {
                 "receiptsRoot": &hash_256_zero,
                 "logsBloom": &bloom_zero,
                 "difficulty": &hash_256_zero,
-                "number": format!("0x{:064x}", 2),
+                "number": &two,
```
Comment (Eikix) [inline src/test_utils/mongo/mod.rs:100]: we can open an issue in the future to create some simple fuzzer for this?

**Review by greged93 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741)* status: 71 of 80 files reviewed, 3 unresolved discussions (waiting on @ClementWalter and @Eikix)

___
*[`src/eth_provider/provider.rs` line 405 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTf21N-TRqi_4U2ayX:-NpVLzYpE6HlVA571ZAs:b-896fix) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L405)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

can we check if there isn't an early return in other clients? (reth?) if block_count > newest_block return default?
</blockquote></details>

Done.
___
*[`src/eth_provider/provider.rs` line 433 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTfTTM5hUhNG2MjFQD:-NpTu818DhpXDKp70j54:b-896fix) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/provider.rs#L433)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

Alright fair
</blockquote></details>

Done.
___
*[`src/eth_provider/provider.rs` line 411 at r5](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpVI68_3tTQRsX2e5U6:-NpVM-De0aYdauL6cciR:bdtqq23) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/645f42b230754924571d0b5a582fa9e8a990014b/src/eth_provider/provider.rs#L411)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

this is a bit weird because in the case of Kakarot, we "consider' that basefee := gasPrice and priorityFee := 0 at all times for compatibility.

TL;DR: we should choose for coherence to use the historical block's basefee per gas using the DB.


1. Use the feeder gateway L1 gasPrice, multiply it by a multiple between 1 and 2 (to account for L2 gas price) for each block and consider it to be the block's baseFee. This is what each user will pay to be included in this block, at a fixed cost.
2. This is now the block's baseFee (for the indexer). Each tx that has a maxFeePerGas below this gasprice will be censored.
3. In the fee history, we can use this base fee as the price of the Kakarot L2.


This gas subject is very tricky indeed! 
- because it is two-dimensional
- because we're trying to be compatible with EIP1559 when we're not

Open to challenges.

It can be a separate macro-task / new issue!
</blockquote></details>

Yeah I am not sure... Any clue what other ZK-EVM's are doing rn? I'm not sure I get your point about the multiplier, L2 gas price is defined by the L1 gas price no?
I think assuming the baseFee == gas price can be ok
___
*[`src/eth_provider/utils.rs` line 19 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTgDKu7toe_ZLZ3doS:-NpTu7W57TNo4Bc7lOed:b-896fix) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/utils.rs#L19)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

I think the comment should be clearer:) in general we should limit the use of "this" , "it" etc. in comments as it might confuse people
</blockquote></details>

Done.
___
*[`src/eth_provider/starknet/kakarot_core.rs` line 64 at r2](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-NpTgypa7LfJio2FXZAj:-NpTu5w97Yxa7My7znq0:b-896fix) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/b7bae11bc93bdf44f4dea8779ad78c7484e39478/src/eth_provider/starknet/kakarot_core.rs#L64)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

we can delete this comment as this is not true anymore
</blockquote></details>

Done.
___
*[`src/test_utils/mongo/mod.rs` line 100 at r5](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#gh-1473123256:-NpVMdGlCUC3Gfop94lT:btscau4) ([raw file](https://github.com/kkrt-labs/kakarot-rpc/blob/645f42b230754924571d0b5a582fa9e8a990014b/src/test_utils/mongo/mod.rs#L100)):*
<details><summary><i>Previously, Eikix (Elias Tazartes) wrote…</i></summary><blockquote>

we can open an issue in the future to create some simple fuzzer for this?
</blockquote></details>

that's complicated, because it means we have to update the tests as well, like for example the test_fee_history return can change based on block numbers.

**Review by Eikix — COMMENTED** [review]
lgtm

> Reviewed 1 of 6 files at r4, 1 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-rpc/741#-:-Np_dMi72PVXryh5YIV3:bb7k0qh)* status: 73 of 80 files reviewed, all discussions resolved (waiting on @ClementWalter)

**Review by Eikix — APPROVED** [review]
lgtm
